File: src/src/main/scala/com/kazzla/asterisk/Session.scala
/*
 * Copyright (c) 2013 koiroha.org.
 * All sources and related resources are available under Apache License 2.0.
 * http://www.apache.org/licenses/LICENSE-2.0.html
*/
package com.kazzla.asterisk

import java.lang.reflect.{Method, InvocationHandler}
import java.util.concurrent.Executor
import java.util.concurrent.atomic.{AtomicReference, AtomicInteger}
import scala.annotation.tailrec
import scala.concurrent.Await
import scala.concurrent.duration.Duration
import org.slf4j.LoggerFactory

// ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
// Session
// ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
/**
 * @author Takami Torao
 * @param name このセッションの名前
 * @param executor このセッション上での RPC 処理を実行するためのスレッドプール
 */
class Session(val name:String, executor:Executor, service:Object, wire:Wire) {

	import Session.logger

	private[this] val pipeIdMask:Short = if(wire.isServer) Pipe.UNIQUE_MASK else 0
	private[this] val pipeSequence = new AtomicInteger(0)
	private[this] val pipes = new AtomicReference[Map[Short, Pipe]](Map())

	@volatile
	private[this] var stub = new Stub(service)

	/**
	 * セッションがクローズされたときに呼び出されるイベントハンドラ。
	 */
	val onClosed = new EventHandlers[Session]()

	/**
	 * このセッション上でピアに提供するサービス (function の集合) を設定。
	 */
	def service_=(service:Object) = stub = new Stub(service)

	// ワイヤーにメッセージが到着した時にメッセージをディスパッチ
	wire.onReceive(m => dispatch(m))

	// ==============================================================================================
	// パイプのオープン
	// ==============================================================================================
	/**
	 * このセッション上で指定された function とのパイプを作成します。
	 * @param function function の識別子
	 * @param params function の実行パラメータ
	 * @return function とのパイプ
	 */
	def open(function:Short, params:AnyRef*):Pipe = {
		val pipe = create(function, params:_*)
		post(Open(pipe.id, function, params:_*))
		pipe
	}

	// ==============================================================================================
	// リモートインターフェースの参照
	// ==============================================================================================
	/**
	 * このセッションの相手側となるインターフェースを参照します。
	 */
	def getRemoteInterface[T](clazz:Class[T]):T = {
		clazz.cast(java.lang.reflect.Proxy.newProxyInstance(
			Thread.currentThread().getContextClassLoader,
			Array(clazz), new Skeleton(clazz)
		))
	}

	/**
	 * 指定されたメッセージを受信したときに呼び出されます。
	 */
	private[this] def dispatch(frame:Message):Unit = {
		if(logger.isTraceEnabled){
			logger.trace(s"-> $frame")
		}
		frame match {
			case open:Open =>
				create(open).foreach{ pipe => stub.call(pipe, open) }
			case close:Close[_] =>
				pipes.get().get(frame.pipeId) match {
					case Some(pipe) => pipe.close(close)
					case None => logger.debug(s"unknown pipe-id: $close")
				}
			case block:Block =>
				pipes.get().get(frame.pipeId) match {
					case Some(pipe) => pipe.receiveQueue.put(block)
					case None => logger.debug(s"unknown pipe-id: $block")
				}
		}
	}

	/**
	 * ピアから指定された Open メッセージを受信したときに呼び出されます。
	 */
	@tailrec
	private[this] def create(open:Open):Option[Pipe] = {
		val map = pipes.get()
		if(map.contains(open.pipeId)){
			post(Close(open.pipeId, null, s"duplicate pipe-id specified: ${open.pipeId}"))
			return None
		}
		val pipe = new Pipe(open.pipeId, this)
		if(pipes.compareAndSet(map, map + (pipe.id -> pipe))){
			return Some(pipe)
		}
		create(open)
	}

	/**
	 * ピアに対して Open メッセージを送信するためのパイプを生成します。
	 */
	@tailrec
	private[this] def create(function:Short, params:AnyRef*):Pipe = {
		val map = pipes.get()
		val id = ((pipeSequence.getAndIncrement & 0x7FFF) | pipeIdMask).toShort
		if(! map.contains(id)){
			val pipe = new Pipe(id, this)
			if(pipes.compareAndSet(map, map + (pipe.id -> pipe))){
				return pipe
			}
		}
		create(function, params:_*)
	}

	/**
	 * 指定されたパイプがクローズされたときに呼び出されます。
	 */
	@tailrec
	private[asterisk] final def destroy(pipeId:Short):Unit = {
		val map = pipes.get()
		if(! pipes.compareAndSet(map, map - pipeId)){
			destroy(pipeId)
		}
	}

	/**
	 * ピアに対して指定されたメッセージを送信します。
	 */
	private[asterisk] def post(frame:Message):Unit = {
		wire.send(frame)
		logger.trace(s"<- $frame")
	}

	/**
	 * このセッションをクローズします。
	 */
	def close():Unit = {
		logger.trace(s"close():$name")
		// TODO 実装まだ

		wire.close()

		onClosed(this)
	}

	// ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	// Stub
	// ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	/**
	 * リフレクションを使用してサービスのメソッド呼び出しを行うためのクラス。
	 */
	private[this] class Stub(service:AnyRef) {

		import scala.language.reflectiveCalls

		logger.debug(s"binding ${service.getClass.getSimpleName} as service of $name")

		private[this] val functions = service.getClass.getInterfaces.map{ i =>
			i.getDeclaredMethods.collect {
				case m if m.getAnnotation(classOf[Export]) != null =>
					val id = m.getAnnotation(classOf[Export]).value()
					logger.debug(s"  function $id to ${m.getSimpleName}")
					id -> m
			}
		}.flatten.toMap

		def call(pipe:Pipe, open:Open):Unit = {
			functions.get(open.function) match {
				case Some(method) =>
					val r = new Runnable {
						def run() {
							Session.sessions.set(Session.this)
							Pipe.pipes.set(pipe)
							try {
								val params = Session.mapParameters(method.getParameterTypes, open.params:_*)
								val result = method.invoke(service, params:_*)
								pipe.close(result)
							} catch {
								case ex:Throwable =>
									logger.debug(s"on call ${method.getSimpleName} with parameter ${open.params}", ex)
									pipe.close(ex)
							} finally {
								Pipe.pipes.remove()
								Session.sessions.remove()
							}
						}
					}
					executor.execute(r)   // TODO アノテーションで別スレッドか同期実行かを選べるようにしたい
				case None =>
					logger.debug(s"unexpected function call: ${open.function} is not defined on class ${service.getClass.getSimpleName}")
					pipe.close(new NoSuchMethodException(open.function.toString))
			}
		}
	}

	// ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	// Skeleton
	// ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	/**
	 * リモート呼び出し先の function を @Export 定義されたメソッドとして扱うための動的プロキシ用ハンドラ。
	 */
	private[this] class Skeleton(clazz:Class[_]) extends InvocationHandler {

		// 指定されたインターフェースのすべてのメソッドに @Export アノテーションが付けられていることを確認
		{
			import scala.language.reflectiveCalls
			val m = clazz.getDeclaredMethods.filter{ m => m.getAnnotation(classOf[Export]) == null }.map{ _.getSimpleName }
			if(m.size > 0){
				throw new IllegalArgumentException(
					s"@${classOf[Export].getSimpleName} annotation is not specified on: ${m.mkString(",")}")
			}
		}

		def invoke(proxy:Any, method:Method, args:Array[AnyRef]):AnyRef = {
			val export = method.getAnnotation(classOf[Export])
			if(export == null){
				// toString() や hashCode() など Object 型のメソッド呼び出し
				method.invoke(this, args:_*)
			} else {
				val pipe = open(export.value(), args:_*)
				val close = Await.result(pipe.future, Duration.Inf)     // TODO アノテーションで呼び出しタイムアウトの設定
				if(close.errorMessage != null){
					throw new Session.RemoteException(close.errorMessage)
				}
				close.result.asInstanceOf[AnyRef]
			}
		}
	}

}

object Session {
	private[Session] val logger = LoggerFactory.getLogger(classOf[Session])

	/**
	 * サービスの呼び出し中にセッションを参照するためのスレッドローカル変数。
	 */
	private[Session] val sessions = new ThreadLocal[Session]()

	// ==============================================================================================
	// セッションの参照
	// ==============================================================================================
	/**
	 * 現在のスレッドを実行しているセッションを参照します。
	 * @return 現在のセッション
	 */
	def apply():Option[Session] = Option(sessions.get())

	class RemoteException(msg:String) extends RuntimeException(msg)

	// ==============================================================================================
	// 型変換
	// ==============================================================================================
	/**
	 * 呼び出しパラメータを指定された型に適切な値へ変換します。
	 * @param types 想定する呼び出しパラメータの型
	 * @param params リモートから渡された呼び出しパラメータ値
	 * @return 呼び出しに使用するパラメータ値
	 */
	private[Session] def mapParameters(types:Array[Class[_]], params:AnyRef*):Array[Object] = {
		types.zip(params).map{ case (t, v) => TypeMapper.appropriateValue(v, t).asInstanceOf[Object] }.toList.toArray
	}

}

File: java8/src/main/java/io/asterisque/Session.java
/*
 * Copyright (c) 2013 koiroha.org.
 * All sources and related resources are available under Apache License 2.0.
 * http://www.apache.org/licenses/LICENSE-2.0.html
*/
package io.asterisque;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.lang.reflect.InvocationHandler;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Function;
import java.util.stream.Stream;

// ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
// Session
// ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
/**
 * ピアとの通信状態を表すクラスです。
 *
 * @author Takami Torao
 */
public class Session extends Attributes{
	private static final Logger logger = LoggerFactory.getLogger(Session.class);

	public final LocalNode node;
	public final String name;
	public final Wire wire;

	/**
	 *
	 * @param node
	 * @param name このセッションの名前
	 * @param defaultService このセッション上でピアに公開する初期状態のサービス
	 * @param wire このセッションのワイヤー
	 */
	Session(LocalNode node, String name, Service defaultService, Wire wire){
		this.node = node;
		this.name = name;
		this.service = defaultService;
		this.wire = wire;
		this.pipeIdMask = wire.isServer()? Pipe.UniqueMask: 0;

		// `Wire` とこのセッションを結合しメッセージポンプを開始
		wire.onClose.add(w -> close());
		/*
		wire.onReceive ++ dispatch        	// `Wire` にメッセージが到着した時のディスパッチャーを設定
		wire.onClosed ++ { _ => close() } 	// `Wire` がクローズされたときにセッションもクローズ
		wire.start()                      	// メッセージポンプを開始
		*/

	}


	/**
	 * このセッションで提供しているサービス。
	 */
	private volatile Service service;

	/**
	 * このセッション上でピアに公開するサービスを変更します。
	 * @param s 新しく公開するサービス
	 * @return 以前のサービス
	 */
	public Service setService(Service s){
		Service old = service;
		service = s;
		return old;
	}

	/**
	 * このセッションで生成するパイプの ID に付加するビットマスク。
	 */
	private final short pipeIdMask;

	/**
	 * 新規のパイプ ID を発行するためのシーケンス番号。
	 */
	private final AtomicInteger pipeSequence = new AtomicInteger(0);

	/**
	 * このセッション上でオープンされているパイプ。
	 */
	private final AtomicReference<Map<Short,Pipe>> pipes = new AtomicReference<>(new HashMap<>());

	/**
	 * このセッションがクローズ処理中かを表すフラグ。
	 */
	private final AtomicBoolean closing = new AtomicBoolean(false);

	/**
	 * このセッションがクローズされているかを表すフラグ。
	 */
	private final AtomicBoolean closed = new AtomicBoolean(false);

	/**
	 * このセッションがクローズされたときに呼び出されるイベントハンドラです。
	 */
	public final EventHandlers<Session> onClosed = new EventHandlers<>();

	// ==============================================================================================
	// パイプのオープン
	// ==============================================================================================
	/**
	 * このセッション上のピアに対して指定された function との非同期呼び出しのためのパイプを作成します。
	 *
	 * @param function function の識別子
	 * @param params function の実行パラメータ
	 * @param onTransferComplete 呼び出し先とのパイプが生成されたときに実行される処理
	 * @return パイプに対する Future
	 */
	public CompletableFuture<Object> open(short function, Object[] params,
																				Function<Pipe,CompletableFuture<Object>> onTransferComplete) {
		Pipe pipe = create(function);
		pipe.open(params);
		CompletableFuture<Object> future = Pipe.using(pipe, () -> onTransferComplete.apply(pipe) );
		pipe.startMessagePump();
		return future;
	}
	public CompletableFuture<Object> open(short function, Object[] params) {
		return open(function, params, pipe -> pipe.future);
	}

	/**
	 * 指定されたメッセージを受信したときに呼び出されその種類によって処理を分岐します。
	 */
	private void dispatch(Message msg){
		if(logger.isTraceEnabled()){
			logger.trace("dispatch: " + msg);
		}
		Pipe pipe;
		if(msg instanceof Open) {
			pipe = create((Open) msg);
		} else {
			pipe = pipes.get().get(msg.pipeId);
		}
		if(pipe != null) {
			if(msg instanceof Open) {
				// サービスを起動しメッセージポンプの開始
				dispatch(pipe, msg);
				pipe.startMessagePump();
			} else {
				// Open は受信したがサービスの処理が完了していない状態でメッセージを受信した場合にパイプのキューに保存できるよう
				// 一度パイプを経由して dispatch(Pipe,Message) を実行する
				pipe.dispatch(msg);
			}
		} else {
			logger.debug("unknown pipe-id: " + msg);
			if(msg instanceof Close){
				logger.debug("both of sessions unknown pipe #" + msg.pipeId);
			} else {
				post(Close.unexpectedError(msg.pipeId, "unknown pipe-id: " + msg.pipeId));
			}
		}
	}

	/**
	 * 指定されたメッセージを受信したときに呼び出されその種類によって処理を分岐します。
	 */
	private void dispatch(Pipe pipe, Message msg) {
		try {
			Pipe.using(pipe, () -> service.dispatch(pipe, msg));
		} catch(Throwable ex){
			logger.error("unexpected error: " + msg + ", closing pipe", ex);
			post(Close.unexpectedError(msg.pipeId, "internal error"));
			if(ex instanceof ThreadDeath){
				throw (ThreadDeath)ex;
			}
		}
	}

	// ==============================================================================================
	// パイプの構築
	// ==============================================================================================
	/**
	 * ピアから受信した Open メッセージに対応するパイプを構築します。オープンに成功した場合は新しく構築されたパイプ
	 * を返します。
	 */
	private Optional<Pipe> create(Open open) {
		while(true){
			Map<Short,Pipe> map = pipes.get();
			// 既に使用されているパイプ ID が指定された場合はエラーとしてすぐ終了
			if(map.containsKey(open.pipeId)){
				logger.debug("duplicate pipe-id specified: " + open.pipeId + "; " + map.get(open.pipeId));
				post(Close.unexpectedError(open.pipeId, "duplicate pipe-id specified: " + open.pipeId));
				return Optional.empty();
			}
			// 新しいパイプを構築して登録
			Pipe pipe = new Pipe(open.pipeId, open.functionId, this);
			Map<Short,Pipe> newMap = new HashMap<>(map);
			newMap.put(pipe.id, pipe);
			if(pipes.compareAndSet(map, newMap)){
				return Optional.of(pipe);
			}
		}
	}

	// ==============================================================================================
	// パイプの構築
	// ==============================================================================================
	/**
	 * ピアに対して Open メッセージを送信するためのパイプを生成します。
	 */
	private Pipe create(short function){
		while(true) {
			Map<Short, Pipe> map = pipes.get();
			short id = (short) ((pipeSequence.getAndIncrement() & 0x7FFF) | pipeIdMask);
			if(!map.containsKey(id)) {
				Pipe pipe = new Pipe(id, function, this);
				Map<Short,Pipe> newMap = new HashMap<>(map);
				newMap.put(pipe.id, pipe);
				if(pipes.compareAndSet(map, newMap)){
					return pipe;
				}
			}
		}
	}

	// ==============================================================================================
	// パイプの破棄
	// ==============================================================================================
	/**
	 * このセッションが保持しているパイプのエントリから該当するパイプを切り離します。
	 */
	void destroy(short pipeId) {
		while(true){
			Map<Short, Pipe> map = pipes.get();
			Map<Short, Pipe> newMap = new HashMap<>(map);
			if(newMap.remove(pipeId) == null || pipes.compareAndSet(map, newMap)){
				return;
			}
		}
	}

	// ==============================================================================================
	// メッセージの送信
	// ==============================================================================================
	/**
	 * ピアに対して指定されたメッセージを送信します。
	 */
	void post(Message msg) throws IOException {
		if(closed.get()){
			throw new IOException("session " + name + " closed");
		} else {
			wire.send(msg);
			if(logger.isTraceEnabled()){
				logger.trace("post: " + msg);
			}
		}
	}

	// ==============================================================================================
	// セッションのクローズ
	// ==============================================================================================
	/**
	 * このセッションをクローズします。
	 * 実行中のすべてのパイプはクローズされ、以後のメッセージ配信は行われなくなります。
	 */
	public void close() {
		if(closing.compareAndSet(false, true)) {
			logger.trace("close(): " + name);

			Map<Short,Pipe> map;
			while(true){
				map = pipes.get();
				if(pipes.compareAndSet(map, Collections.emptyMap())){
					break;
				}
			}

			// 残っているすべてのパイプに Close メッセージを送信
			map.values().forEach( pipe -> {
				pipe.close(new Close(pipe.id, new Abort(Abort.SessionClosing, "session " + name + " closing")));
			});

			post(new Control(Control.Close, new byte[0]));

			// 以降のメッセージ送信をすべて例外に変更して送信を終了
			// ※Pipe#close() で Session#post() が呼び出されるためすべてのパイプに Close を投げた後に行う
			closed.set(true);

			// Wire のクローズ
			wire.close();

			// セッションのクローズを通知
			onClosed.accept(this);
		}
	}

	// ==============================================================================================
	// リモートインターフェースの参照
	// ==============================================================================================
	/**
	 * このセッションの相手側となるインターフェースを参照します。
	 */
	public <T> T bind(Class<T> clazz){
		return clazz.cast(java.lang.reflect.Proxy.newProxyInstance(
			Thread.currentThread().getContextClassLoader(),
			new Class[]{ clazz }, new Skeleton(clazz)
		));
	}

	// ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	// Skeleton
	// ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	/**
	 * リモート呼び出し先の function を @Export 定義されたメソッドとして扱うための動的プロキシ用ハンドラ。
	 */
	private class Skeleton implements InvocationHandler {

		public Skeleton(Class<?> clazz){
			// 指定されたインターフェースのすべてのメソッドに @Export アノテーションが付けられていることを確認
			Optional<String> methods = Stream.of(clazz.getDeclaredMethods())
				.filter(m -> m.getAnnotation(Export.class) == null)
				.map(Debug::getSimpleName).reduce((a, b) -> a + "," + b);
			if(methods.isPresent()){
				throw new IllegalArgumentException(
					"@" + Export.class.getSimpleName() + " annotation is not specified on: " + methods.get());
			}
		}

		// ============================================================================================
		// リモートメソッドの呼び出し
		// ============================================================================================
		/**
		 * リモートメソッドを呼び出します。
		 * @param proxy プロキシオブジェクト
		 * @param method 呼び出し対象のメソッド
		 * @param args メソッドの引数
		 * @return 返し値
		 */
		public Object invoke(Object proxy, Method method, Object[] args) throws InvocationTargetException, IllegalAccessException {
			Export export = method.getAnnotation(Export.class);
			if(export == null){
				// toString() や hashCode() など Object 型のメソッド呼び出し?
				logger.debug("normal method: " + Debug.getSimpleName(method));
				return method.invoke(this, args);
			} else {
				// there is no way to receive block in interface binding
				return open(export.value(), (args==null? new Object[0]: args));
			}
		}

	}

}
